homeassistant/components/sentry/__init__.py

```diff
@@ -44,15 +51,98 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    custom_components = await async_get_custom_components(hass)
+
+    def before_send(event, hint):
+        # Filter out handled events by default
+        if "tags" in event and event.tags.get("handled", "no") == "yes":
+            return None
+
+        # Additional tags to add to the event
+        additional_tags = {
+            "channel": channel,
+            "installation_type": system_info["installation_type"],
+            "uuid": huuid,
+        }
+
+        # Find out all integrations in use, filter "auth", because it
+        # triggers security rules, hiding all data.
+        integrations = [
+            integration
+            for integration in hass.config.components
+            if integration != "auth" and "." not in integration
+        ]
+
+        # Additional extra data to add to the event
+        extra_data = {
+            **system_info,
+            "custom_components": "\n".join(sorted(custom_components)),
+            "integrations": "\n".join(sorted(integrations)),
+        }
+
+        # Add additional tags based on what caused the event.
+        platform = entity_platform.current_platform.get()
+        if platform is not None:
+            # This event happened in a platform
+            additional_tags["custom_component"] = "no"
+            additional_tags["integration"] = platform.platform_name
+            additional_tags["platform"] = platform.domain
+        elif "logger" in event:
+            # Logger event, try to get integration information from the logger name.
+            matches = LOGGER_INFO_REGEX.findall(event["logger"])
+            if matches:
+                group1, group2, group3, group4 = matches[0]
+                # Handle the "homeassistant." package differently
+                if group1 == "homeassistant" and group2 and group3:
+                    if group2 == "components":
+                        # This logger is from a component
+                        additional_tags["custom_component"] = "no"
+                        additional_tags["integration"] = group3
+                        if group4 and group4 in ENTITY_COMPONENTS:
+                            additional_tags["platform"] = group4
+                    else:
+                        # Not a component, could be helper, or something else.
+                        additional_tags[group2] = group3
+                else:
+                    # Not the "homeassistant" package, this third-party
+                    additional_tags["package"] = group1
+
+        # If this event is caused by an integration, add a tag if this
+        # integration is custom or not.
+        if (
+            "integration" in additional_tags
+            and additional_tags["integration"] in custom_components
+        ):
+            additional_tags["custom_component"] = "yes"
+
+        # Update event with all additional data
+        event.setdefault("tags", {}).update(additional_tags)
+        event.setdefault("extra", {}).update(extra_data)
```

Comment:
one thing that may be valuable is using the 'contexts' api here. i'm not sure if its present by default but basically:

event.setdefault("contexts", {}).setdefault("homeassistant", extra_day)

While this doesnt do anything mind blowing today, it will:

1. rendering it its own box (and avoid squashing anything in 'extra'
2. eventually be (hopefully) fully indexed/searchable

one thing to note if you use that api is don't use the "type" param (right now it wont really do anything, but its semi-reserved). if you want you an explicitly set ``"type": "default"`` as well.

see https://github.com/getsentry/sentry/blob/master/src/sentry/interfaces/contexts.py for general implementation - though most is abstracted in our Rust bindings these days.